File: 0018-4sum/0018-4sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        return kSum(nums, 0, 4, (long) target);
    }

    private List<List<Integer>> kSum(int[] nums, int start, int k, long target) {
        List<List<Integer>> res = new ArrayList<>();

        if (k == 2) {
            int l = start, r = nums.length - 1;
            while (l < r) {
                long sum = (long) nums[l] + nums[r];
                if (sum == target) {
                    res.add(Arrays.asList(nums[l], nums[r]));
                    l++; r--;
                    while (l < r && nums[l] == nums[l - 1]) l++;
                    while (l < r && nums[r] == nums[r + 1]) r--;
                } else if (sum < target) {
                    l++;
                } else {
                    r--;
                }
            }
            return res;
        }

        for (int i = start; i <= nums.length - k; i++) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            for (List<Integer> subset : kSum(nums, i + 1, k - 1, target - nums[i])) {
                ArrayList<Integer> cur = new ArrayList<>();
                cur.add(nums[i]);
                cur.addAll(subset);
                res.add(cur);
            }
        }
        return res;
    }
}

File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = set()
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                for k in range(j + 1, len(nums)):
                    for l in range(k + 1, len(nums)):
                        if nums[i] + nums[j] + nums[k] + nums[l] == target:
                            temp = [nums[i], nums[j], nums[k], nums[l]]
                            temp.sort()
                            res.add(tuple(temp))

        return [list(x) for x in res]
